File: dichotomy.m
%二分法
clear;
clc;
syms U L;    %将区间上下限定为变量
f=@(x) sin(x)+x+1;    %求给定的函数，可以直接在本行中修改后面代码为其他函数
U=input('输入求根区域上限upper,U=');    
L=input('输入求根区域下限lower,L=');
while U-L>0.000001    %设定一个求根区域精度，然后进行判断
    root=(U+L)/2;    %当根的区间大于所给精度时，利用二分法重新规划求根区间
    if f(root)==0    
        break;    %r恰好为所求根，直接跳出循环
    end
    if f(root)*f(U)<0    %用零点存在定理判断根所在的区域
        L=root;
    else
        U=root;
    end
end
root    %直接输出所求根的值

 

File: C++代码/dichotomy/dichotomy.cpp
// dichotomy.cpp : 此文件包含 "main" 函数。程序执行将在此处开始并结束。
//

#include "pch.h"
#include <iostream>
using namespace std;
double fun(double x);
int main()
{
	double L, U, e, result;
	cout << "请输入求根区域下限lower,L=";
	cin >> L;
	cout << "请输入求根区域上限upper,U=";
	cin >> U;
	cout << "请输入误差限error,e=";
	cin >> e;

	while (abs(U - L) > e) {
		double root = (U + L) / 2.0;
		//double y = fun(root);
		if (fun(root) == 0)
			break;
		if (fun(root) * fun(U) > 0)
			U = root;
		else
			L = root;
		result = root;
	}
	cout << "方程的根为" << result << endl;
}

double fun(double x) {
	double y = sin(x) + x + 1;
	return y;
}
